File: src/ts/examples/to-cityjson.ts
/** The CityJSON representation, and how to reach into its fields.
 *
 *      node examples/to-cityjson.ts in.fcb [featureIndex]
 *
 *  `toCityJSONMetadata` gives the CityJSONSeq header line and
 *  `toCityJSONFeature` gives one feature line. Both return plain typed
 *  objects, so everything below is ordinary property access.
 */
import { toCityJSONFeature, toCityJSONMetadata } from '@cityjson/flatcitybuf'
import { fromFile } from '@cityjson/flatcitybuf/node'

const path = process.argv[2]
const index = process.argv[3] === undefined ? 0 : Number(process.argv[3])
if (path === undefined) {
  console.log('usage: node examples/to-cityjson.ts <file.fcb> [featureIndex]')
  process.exit(2)
}

const reader = await fromFile(path)
try {
  console.log('== metadata (toCityJSONMetadata) ==')
  const meta = toCityJSONMetadata(reader.header)
  console.log(`  version   ${meta.version}`)
  console.log(`  scale     [${meta.transform.scale.join(', ')}]`)
  console.log(`  translate [${meta.transform.translate.join(', ')}]`)
  if (meta.metadata?.referenceSystem !== undefined) {
    console.log(`  CRS       ${meta.metadata.referenceSystem}`)
  }
  if (meta.metadata?.geographicalExtent !== undefined) {
    console.log(`  extent    [${meta.metadata.geographicalExtent.join(', ')}]`)
  }

  // Present only when the file has them. The templates' material and
  // texture mappings index the header's OWN appearance palette, which is
  // why both must be emitted together -- emit one without the other and
  // those mappings point at nothing.
  const templates = meta['geometry-templates']
  if (templates !== undefined) console.log(`  templates ${templates.templates.length}`)
  if (meta.appearance !== undefined) {
    const ap = meta.appearance
    console.log(
      `  palette   ${ap.materials?.length ?? 0} material(s), ` +
        `${ap.textures?.length ?? 0} texture(s)`,
    )
  }
  if (meta.extensions !== undefined) {
    console.log(`  extensions ${Object.keys(meta.extensions).sort().join(', ')}`)
  }

  console.log()
  console.log(`== feature ${index} (toCityJSONFeature) ==`)
  let cj
  for await (const feature of await reader.select({ offset: index, limit: 1 })) {
    cj = toCityJSONFeature(feature, reader.header)
  }
  if (cj === undefined) {
    console.error(`no feature at index ${index}`)
    process.exit(1)
  }

  console.log(`  id        ${cj.id}`)
  console.log(`  vertices  ${cj.vertices.length}`)
  for (const [objId, obj] of Object.entries(cj.CityObjects)) {
    const geoms = obj.geometry ?? []
    console.log(`  object    ${objId}`)
    console.log(`    type      ${obj.type}`)
    console.log(`    geometry  ${geoms.length} (lod ${geoms.map((g) => g.lod).join(', ')})`)
    const attrs = obj.attributes
    if (attrs !== undefined && Object.keys(attrs).length > 0) {
      const keys = Object.keys(attrs)
      const head = keys.slice(0, 3).map((k) => `${k}=${JSON.stringify(attrs[k])}`)
      console.log(`    attrs     ${keys.length}, e.g. ${head.join(' ')}`)
    }
  }

  console.log()
  const bytes = JSON.stringify(cj).length
  console.log('  the whole feature as one JSON line is what read-local.ts')
  console.log(`  writes; it is ${bytes} bytes here`)
} finally {
  await reader.close()
}

File: src/py/examples/to_cityjson.py
"""The CityJSON representation, and how to reach into its fields.

    python examples/to_cityjson.py in.fcb [feature_index]

`to_cityjson_metadata` gives the CityJSONSeq header line and
`to_cityjson_feature` gives one feature line; both return plain dicts,
so everything below is ordinary dict access, not a bespoke API.
"""

from __future__ import annotations

import itertools
import json
import sys

import flatcitybuf as fcb


def main(path: str, index: int) -> int:
    reader = fcb.FcbReader.open_file(path)

    print("== metadata (to_cityjson_metadata) ==")
    meta = fcb.to_cityjson_metadata(reader.header)
    print(f"  version   {meta['version']}")
    tr = meta["transform"]
    print(f"  scale     {tr['scale']}")
    print(f"  translate {tr['translate']}")
    md = meta.get("metadata", {})
    if "referenceSystem" in md:
        print(f"  CRS       {md['referenceSystem']}")
    if "geographicalExtent" in md:
        print(f"  extent    {md['geographicalExtent']}")

    # Present only when the file has them. The templates' material and
    # texture mappings index the header's OWN appearance palette, which
    # is why both must be emitted together.
    if "geometry-templates" in meta:
        t = meta["geometry-templates"]["templates"]
        print(f"  templates {len(t)}")
    if "appearance" in meta:
        ap = meta["appearance"]
        print(
            f"  palette   {len(ap.get('materials', []))} material(s), "
            f"{len(ap.get('textures', []))} texture(s)"
        )
    if "extensions" in meta:
        print(f"  extensions {sorted(meta['extensions'])}")

    print()
    print(f"== feature {index} (to_cityjson_feature) ==")
    feature = next(itertools.islice(reader.select_all(), index, index + 1))
    cj = fcb.to_cityjson_feature(feature, reader.header)
    print(f"  id        {cj['id']}")
    print(f"  vertices  {len(cj['vertices'])}")
    for obj_id, obj in cj["CityObjects"].items():
        geoms = obj.get("geometry", [])
        lods = [g.get("lod") for g in geoms]
        print(f"  object    {obj_id}")
        print(f"    type      {obj['type']}")
        print(f"    geometry  {len(geoms)} (lod {lods})")
        attrs = obj.get("attributes", {})
        if attrs:
            first = dict(itertools.islice(attrs.items(), 3))
            print(f"    attrs     {len(attrs)}, e.g. {first}")

    print()
    print("  the whole feature as one JSON line is what read_local.py")
    print(f"  writes; it is {len(json.dumps(cj))} bytes here")
    return 0


if __name__ == "__main__":
    if len(sys.argv) not in (2, 3):
        print(__doc__)
        raise SystemExit(2)
    i = int(sys.argv[2]) if len(sys.argv) == 3 else 0
    raise SystemExit(main(sys.argv[1], i))
